common/psbt_open.c

```diff
@@ -494,20 +492,26 @@ bool psbt_output_to_external(const struct wally_psbt_output *output)
 	return !(!result);
 }
 
-bool psbt_contribs_changed(struct wally_psbt *orig,
+bool psbt_contribs_changed(const struct wally_psbt *orig,
 			   struct wally_psbt *new)
 {
-	assert(orig->version == 2 && new->version == 2);
-
+	struct wally_psbt *orig_deep_copy;
 	struct psbt_changeset *cs;
 	bool ok;
-	cs = psbt_get_changeset(NULL, orig, new);
+
+	assert(orig->version == 2 && new->version == 2);
+	// The `psbt_get_changeset` function modifies the original PSBT, which
+	// is not needed by the .
+	//
+	// This behavior also caused issues with a different hsmd, as documented in:
+	// https://github.com/ElementsProject/lightning/issues/6672
+	orig_deep_copy = clone_psbt(tmpctx, orig);
+	cs = psbt_get_changeset(tmpctx, orig_deep_copy, new);
```

Comment:
better fix would be to update `linearize_input` so that it doesn't mutate the inputs; that'll fix it for all calling paths rather than this one.

see
https://github.com/ElementsProject/lightning/issues/6672#issuecomment-1756461897

Reply:
@niftynei looking at `linearize_input` on `master` - https://github.com/ElementsProject/lightning/blob/e1ac2410b01dfcf17e806418c6d01c336c79a78c/common/psbt_open.c#L66-L74

I see:

- `psbt->inputs[0] = *in;` - that makes a copy of the input struct bytes, copying any pointers without copying underlying memory
- `wally_psbt_input_set_final_witness(&psbt->inputs[0], NULL);` - this uses the `SET_STRUCT` macro, which in turn **frees the memory originally pointed to**.  as far as I understand, this results in a dangling pointer in the original input struct.
- there's also the `/* Hide the inputs we added` comment below, but it's too late by that time

so I'm actually not sure why the code in `master` doesn't just crash.

but in any case, a potential fix to this and the unwanted mutation may be to do `psbt->inputs[0].final_witness = NULL`, so the original memory is not touched.  and the same for the other four calls of the form `wally_psbt_input_set_*()`